wallet: batch all individual spkms setup db writes in a single db txn

furszy opened the pull request:
Work decoupled from #28574.

Instead of performing multiple single write operations per spkm
setup call, this PR batches them all within a single atomic db txn.

Speeding up the process and preventing the wallet from entering
an inconsistent state if any of the intermediate transactions fail
(which shouldn't happen but.. if it does, it is better to not store
any spkm rather than storing them partially).

To compare the changes, added benchmark in the first commit.

furszy committed:
wallet: batch legacy spkm TopUp

Instead of performing multiple atomic write
operations per legacy spkm setup call, batch
them all within a single atomic db txn.

furszy committed:
wallet: batch descriptor spkm TopUp

Instead of performing multiple atomic write
operations per descriptor setup call, batch
them all within a single atomic db txn.

furszy committed:
wallet: descriptors setup, batch db operations

Instead of doing one db transaction per descriptor setup,
batch all descriptors' setup writes in a single db txn.

Speeding up the process and preventing the wallet from entering
an inconsistent state if any of the intermediate transactions
fail.

furszy committed:
bench: add benchmark for wallet creation procedure

theStack reviewed (commented):
Concept ACK

Didn't look deeper at the code yet, but ran the new benchmark (on an SSD) for comparison via
```
$ ./src/bench/bench_bitcoin -filter=WalletCreatePlain\|WalletCreateEncrypted
```

master (commit bb4554c81e0d819d74996f89cbb9c00476aedf8c which adds the benchmark)                                                
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      797,011,189.00 |                1.25 |    0.6% |      8.74 | `WalletCreateEncrypted`
|      391,984,469.00 |                2.55 |    1.1% |      4.30 | `WalletCreatePlain`

PR head (1901360ca52c6563a40dc78633cc492a822ac7a9):
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      724,523,571.00 |                1.38 |    0.5% |      7.97 | `WalletCreateEncrypted`
|      319,389,684.00 |                3.13 |    0.8% |      3.50 | `WalletCreatePlain`

Looks like a rough ~10% speedup for encrypted and ~22% speedup for plain wallets on my machine. Even without speedups, I think it's good practice to avoid entering an inconsistent state, as mentioned in the PR description.

Sjors commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -582,6 +582,9 @@ class DescriptorScriptPubKeyMan : public ScriptPubKeyMan
     std::unique_ptr<FlatSigningProvider> GetSigningProvider(int32_t index, bool include_private = false) const EXCLUSIVE_LOCKS_REQUIRED(cs_desc_man);
 
+    //! Same as 'TopUp' but designed for use within a batch transaction context
+    bool TopUpWithDb(WalletBatch& batch, unsigned int size = 0);
```
da133a80bae311373337a0d9c474667a5f73908d nit: `TopUpWithDB`

furszy replied:
Done as suggested

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3535,23 +3535,30 @@ void CWallet::SetupDescriptorScriptPubKeyMans(const CExtKey& master_key)
         }
     }
+
+    // Ensure information is dumped to disk
```
1901360ca52c6563a40dc78633cc492a822ac7a9: nit "is committed to disk"

furszy replied:
Done as suggested

Sjors reviewed (approved):
ACK 1901360ca52c6563a40dc78633cc492a822ac7a9

Not sure if b85b75d6470841813c5844b2e9c883360f1ad94a is needed since we're getting rid of the legacy wallet anyway. But it looks correct.

Ran bench on MacBook Pro 2019 (2,3 GHz 8-Core Intel Core i9, SSD):

Before:

```
src/bench/bench_bitcoin -filter="WalletCreate[E|P].*" -min-time=30000                          
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      924,961,364.67 |                1.08 |    1.8% |     30.41 | `WalletCreateEncrypted`
|      537,964,679.50 |                1.86 |    2.3% |     35.03 | `WalletCreatePlain`

After:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      777,090,017.50 |                1.29 |    1.3% |     32.65 | `WalletCreateEncrypted`
|      361,932,639.25 |                2.76 |    0.6% |     33.34 | `WalletCreatePlain`

Here's a commit that batches the descriptor import for external signers: https://github.com/Sjors/bitcoin/commit/286883a68d24e3600d9e608ccdea55763c50dde1 (I had to make `TopupWithDB()` public)

Sjors committed:
wallet: batch external signer descriptor import

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

furszy commented:
Updated per feedback, thanks for the review @Sjors.

> Here's a commit that batches the descriptor import for external signers: https://github.com/Sjors/bitcoin/commit/286883a68d24e3600d9e608ccdea55763c50dde1 (I had to make TopupWithDB() public)

Thanks. Pulled the commit and made two changes to it:
1) Instead of making `TopupWithDB()` public, made it protected. The `ExternalSignerScriptPubKeyMan` class is derived from `DescriptorScriptPubKeyMan`.
2) Removed the unimplemented `SetupDescriptor()` function from `DescriptorScriptPubKeyMan`. Only `ExternalSignerScriptPubKeyMan` utilizes it.

Sjors commented:
re-utACK f05302427386fe63f4929a7198652cb1e4ab3bcc

CI failure seems spurious

S3RK commented:
Concept and Approach ACK

achow101 commented:
ACK f05302427386fe63f4929a7198652cb1e4ab3bcc

BrandonOdiwuor reviewed (approved):
ACK f05302427386fe63f4929a7198652cb1e4ab3bcc

theStack reviewed (approved):
Code-review ACK f05302427386fe63f4929a7198652cb1e4ab3bcc
